### DISCUSSION TITLE
What is the difference between arm v5a and v5t ?

### DISCUSSION THREAD

**Discussion by csersoft** [upx/upx#692] [Q&A] [answered]
This issue tracker is **ONLY** used for requesting sensible features.
Please use [stackoverflow](https://stackoverflow.com) for supporting issues.

## What's the problem (or question)?
There are two files about arm shared lib in the stub directory, which are **arm.v5a-linux.shlib-init.h** and **arm.v5t-linux.shlib-init.h**. The **v5a** file seems to be useless. What is their difference please?

## What should have happened?
No

## Do you have an idea for a solution?
No

## Please tell us details about your environment.

* UPX version used (`upx --version`): 4.02
* Host Operating System and version: Ubuntu 20.04
* Host CPU architecture: x86_64
* Target Operating System and version:
* Target CPU architecture:

**Comment by jreiser** [comment, marked as answer]
v5t hardware has Thumb instructions (generally 16-bit) which usually enable smaller total code size, especially for code which is control-dominated (many `if-else`, short `then`, etc.) v5a hardware does not have Thumb instructions, and allows "equal access" to registers r8 through r15, so tends to be better for "data-dominated" code: encryption, heavy use of arrays, etc.  However high-numbered registers already have reserved uses: pc==r15, lr==r14, sp==r13, linker_fixup==r12, fp==r11 (or r10?); so there aren't that many more registers available for general use than the r0-r7 of Thumb mode.  (And Thumb mode can access high registers for add/subtract/compare/move.)
Also, later hardware has `bx` and `blx` (branch with target pointed to by arbitrary register) which earlier hardware does not.
And v4 is really ancient, with many fewer instances surviving thatn other generations.
Currently the upx de-compression sub for shared library is invoked in Thumb mode always, and nobody has complained.

**Reply by csersoft** [reply]
Thanks

**Comment by jreiser** [comment]
The original plan was to have different de-compressors for v5a and v5t.  But the population of devices seemed largely skewed towards v5t.  And then ARM slapped developers in the face, because the only way to tell which hardware an executable assumes (requires) is to parse and decode the ARM_ATTRIBUTES ElfXX_Shdr section.  This is not nice because it is unlike any other $(ARCH), and the documentation is arcane, and it requires dozens of lines of code.  ARM should have done what MIPS did: add DT_xxxxx entries (see /usr/include/elf.h for DT_MIPS_xxxxx) to the PT_DYNAMIC segment, so that the parsing and decoding is trivial, and the info is in PT_DYNAMIC segment, the Execution view, where it belongs, instead of the ElXXx_Shdr Section, which is the Linking view.  By design of ELF, the OS can pay attention only to ElfXX_Phdrs, and totally ignore ElfXX_Shdrs.  (So upx can just discard Shdrs.)  But ARM ignored that very useful distinction. Android went even further in thumbing their nose at the ELF specification, by groveling into every nook and cranny of ElfXX_Shdrs.

And then hardware-wise, years of ARM hardware was unfit for co-operating multi-processing.  Semaphores, mutual exclusion, etc., was essentially impossible on ARM v4 and v5.  On ARM v6, it was possible but very ugly and slow: more than a dozen instructions. Finally ARM v7 was fit for use by modern multiprocessing apps.  Today armv7 is the minimum supported by nearly all Linux distributions, and some are even migrating to armv8 (aarch64) as the minimum.
